spatialos-sdk-tools/src/setup/main.rs

```diff
@@ -0,0 +1,201 @@
+        for entry in glob::glob(schema_glob.to_str().unwrap())?.filter_map(Result::ok) {
+            command.arg(&entry);
+        }
+    }
+
+    trace!("{:#?}", command);
+    command
+        .status()
+        .map_err(|_| "Failed to compile schema files")?;
+
+    // Run protoc on all the generated proto files.
+    let proto_path_arg = OsString::from("--proto_path=").tap(|arg| arg.push(&tmp_path));
+    let mut command = Command::new(&protoc_path);
+    command
+        .arg(&proto_path_arg)
+        .arg(&descriptor_out_arg)
+        .arg("--include_imports");
+    let proto_glob = tmp_path.join("**/*.proto");
+    for entry in glob::glob(proto_glob.to_str().unwrap())?.filter_map(Result::ok) {
+        command.arg(&entry);
+    }
+
+    trace!("{:#?}", command);
+    command.status().map_err(|_| "Failed to run protoc")?;
+
+    // Remove the temp directory once the setup process has finished.
+    fs::remove_dir_all(&tmp_path).map_err(|_| "Failed to remove temp dir")?;
+
+    // If the user specified the `--codegen` flag, run code generation with the bundle file
+    // that we just generated.
+    if let Some(codegen_out_path) = codegen {
+        let mut input_file =
+            File::open(&bundle_json_path).expect("Unable to open the test schema bundle.");
+        let mut contents = String::new();
+        input_file
+            .read_to_string(&mut contents)
+            .expect("Unable to read the test schema bundle");
+        let generated_file =
+            generator::generate_code(schema_bundle::load_bundle(&contents).unwrap());
+        let mut output_file = File::create(codegen_out_path).unwrap();
+        output_file.write_all(generated_file.as_bytes()).unwrap();
+    }
+
+    Ok(())
+}
+
+#[derive(Debug, StructOpt)]
+#[structopt(
+    name = "setup",
+    rename_all = "kebab-case",
+    about = "Perform setup for a Rust SpatialOS project."
+)]
+struct Opt {
+    /// The path to your local installation of the SpatialOS SDK
+    ///
+    /// If not specified, uses the SPATIAL_OS_DIR environment variable instead. Will fail
+    /// with an error if neither is set.
+    #[structopt(long, short = "l", parse(from_os_str))]
+    spatial_lib_dir: Option<PathBuf>,
+
+    /// A directory to search for schema files
+    ///
+    /// The directory will be searched recursively for all .schema files. Any schema
+    /// files found will be included in compilation. Can be specified multiple times,
+    /// e.g. `setup -s foo/schemas -s bar/schemas -o schemas/bin`.
+    #[structopt(long = "schema-path", short = "s", parse(from_os_str))]
+    schema_paths: Vec<PathBuf>,
+
+    /// Display detailed log output
+    #[structopt(long, short)]
+    verbose: bool,
+
+    /// Perform code generation and put the output in the specified file
+    ///
+    /// If not specified, will not perform code generation.
+    #[structopt(long, short = "c", parse(from_os_str))]
+    codegen: Option<PathBuf>,
```

Comment:
Hrmmm, so I originally had it as something like that, but changed it to `--codegen` because I wanted it to more clearly indicate that specifying that flag was also enabling codegen. I can change it back to `--codegen-out-path` if you think that's better, I don't have super strong feelings about it either way. Though, if you think `--codegen` is fine as the CLI flag and you just want the variable to be `codegen_out_path`, I can also do that 😁